src/cargo/util/toml/mod.rs

```diff
@@ -880,11 +964,74 @@ where
         where
             E: de::Error,
         {
-            string.trim().parse().map_err(de::Error::custom)
+            match string.trim().parse().map_err(de::Error::custom) {
+                Ok(parsed) => Ok(MaybeWorkspace::Defined(parsed)),
+                Err(e) => Err(e),
+            }
+        }
+
+        fn visit_map<V>(self, map: V) -> Result<Self::Value, V::Error>
+        where
+            V: de::MapAccess<'de>,
+        {
+            let mvd = de::value::MapAccessDeserializer::new(map);
+            TomlWorkspaceField::deserialize(mvd).map(MaybeWorkspace::Workspace)
+        }
+    }
+
+    deserializer.deserialize_any(Visitor)
+}
+
+/// Enum that allows for the parsing of { workspace = true } in a Cargo.toml
+///
+/// It allows for things to be inherited from a workspace or defined as needed
+#[derive(Deserialize, Serialize, Clone, Debug)]
+#[serde(untagged)]
+pub enum MaybeWorkspace<T> {
+    Workspace(TomlWorkspaceField),
+    Defined(T),
+}
+
+impl<T> MaybeWorkspace<T> {
+    fn resolve<'a>(
+        self,
+        cargo_features: &Features,
+        label: &str,
+        inherit: impl FnOnce() -> Option<&'a InheritableFields>,
+        get_ws: impl FnOnce(&InheritableFields) -> Option<T>,
```

Comment:
Why do we need both instead of just `impl FnOnce() -> Option<T>`?

If we switch it to `impl FnOnce( -> CargoResult<T>` then could we get rid of the `label` parameter?  I guess it would make the error less nice for the `workspace = false` case.  Unsure on the trade off there.

Reply:
The [original comment](https://github.com/rust-lang/cargo/pull/9684#discussion_r728453774) had a `label` and was `get_ws: impl FnOnce() -> &T`. I modified it to what it is now for convenience but I think no matter what we should leave `label` since it makes the error messages nicer in the current implementation.

As for going to `impl FnOnce() -> Option<T>`, I will look into this and report back. If I remember to when I changed it to this, it made things a bit easier to understand and split getting the fields to inherit from and how to get it from those inheritable fields. I can see why one might be a better implementation since that way if we already have inheritable fields we could directly call in without `inherit: impl FnOnce() -> Option<&'a InheritableFields>,`

Reply:
>  but I think no matter what we should leave label since it makes the error messages nicer in the current implementation.

This is why I was wondering about the signature `impl FnOnce( -> CargoResult<T> ` as it could do the error formatting as well.  Granted, that makes it so we duplicate the error messages which makes it more likely for them to be unintentionally different, so maybe keeping it an `Option` with `label` is better.